tuist/tuist
# Add Workspace Mappers

Part of #1765 

### Short description 📝

Recently as part of #1765 we identified a use case that required the introduction of a workspace wide mapper to apply modification to the top level workspace to enable the creation of new schemes.

### Solution 📦

Introduce `WorkspaceMapper`s that can be composed with the existing `ProjectMapper`s to enable the creation of a `AutomaticSchemesWorkspaceMapper` in #1765. 

### Implementation 👩‍💻👨‍💻

- [x] Introduce `WorkspaceMapper` + `ProjectWorkspaceMapper`
- [x] Integrate `WorkspaceMapper`s into the generation pipeline 

### Test Plan 🛠

Verify existing mappers continue to function _(most of this is already covered by acceptance tests)_, for example:

- Run `swift run tuist generate` within `fixtures/ios_app_with_custom_scheme`
- Verify workspace & projects are generated successfully
- Verify automaticly generates schemes (e.g. one per Target) still appear
- Run `swift run tuist generate` within `fixtures/ios_app_with_framework_and_resources`
- Verify workspace & projects are generates successfully
- Verify resouce accessors are still generated (`Assets+App.swift`)
